### ISSUE TITLE
ELEC-1 — Electrostatics extra: `graph_electrostatics` reference solver, solver-dispatch layer, and the expert training config section

### ISSUE THREAD

**Issue by aacostadiaz** [ACEsuit/mace#1591]
**Depends on:** INF-1 (#1550), ARCH-3 (#1562), TRN-1 (#1575) · **Blocks:** ELEC-2 (#1592), DEP-1a (#1634), GATE-3 (#1585), ELEC-4 (#1635)

**Context:** Long-range electrostatics enters v1 as an **optional extra**, not core. Today the two electrostatics models — `MACELES` and `PolarMACE` — live in `mace/modules/extensions.py:142,666` (a 2105-line file), where both **subclass `ScaleShiftMACE` and override `forward`**; `PolarMACE` adds k-space electrostatic energies that depend on positions and cell and so couple into the derivative engine. That means electrostatics is **models, not readouts**: the plan's "extending touches zero core files" property (true for backends, observables, losses, data formats, readouts) does **not** apply here. A new electrostatics model in v1 is a `BaseMACE` subclass + a `MACEOutputs` (ARCH-3 (#1562)) + a **model-transform hook** (TRN-1 (#1575)'s staged pipeline), never an observable.

Underneath those models sits the **long-range / k-space solver** (Ewald/PME sum, plus the SCF fixed point that charge-aware models close on top). That solver is itself backend-swappable: a plain-torch reference (Will Baldwin's `graph_electrostatics`, **import name `graph_longrange`**, which `mace/modules/extensions.py:14-24` imports inside a `try/except` that sets `GRAPH_LONGRANGE_AVAILABLE`) versus accelerated libraries (NVIDIA `nvalchemiops`, prototyped on `upstream/agent/nvalchemiops-polar-backend`). This ticket lands the **reference solver + the solver-dispatch layer + the `[electrostatics]` extra + a paths-filtered CI job**, so ELEC-2 (#1592) (PolarMACE) and ELEC-4 (#1635) (LES) are writable on top. Neither model is in scope here.

Two facts about today's packaging bind this ticket. **There is no `[electrostatics]` extra on develop**: the solver is a *pinned git dependency* in `requirements/polar.txt` (`git+https://github.com/WillBaldwin0/graph_electrostatics.git@0e21d554`), installed by the `polar` / `polar-cueq` / `torchsim` jobs in `ci-extensions.yaml` via the `setup-mace` action's `pip-packages` input, and probed by the `polar` capability marker as `_module_available("graph_longrange")` (`tests/conftest.py:64`). And **develop has no runtime solver selector at all** — no `set_electrostatics_backend`, no `--electrostatics_backend`, no calculator argument. The dispatch layer is therefore new machinery, not a rework of an existing leak; what it must avoid is *reintroducing* the runtime selection the kernel-backend rewrite removed for the message-passing ops.

**Interface & constraints:**

The solver-dispatch layer mirrors the kernel-backend dispatch (BKD-1 (#1560)) **one op class over**, and is kept in a **separate registry** so the message-passing op contract stays closed.

1. **Descriptor (framework-free, in `mace_core`, dtypes are names).** `ElectrostaticsSolverDescriptor` carries everything a solver needs to decide whether it can run:

| Field | Meaning |
|---|---|
| `periodicity_profile` | one of `{full_periodic, z_slab, molecular, partial}` |
| `slab_normal` | axis for `z_slab` corrections |
| `multipole_max_l` | multipole/expansion order |
| `kspace_cutoff`, `smearing_width` | k-space sum parameters |
| `scf_spec` | `{max_iters, tolerance, mixing}` (absent for non-SCF models) |
| `external_field_flags` | applied-field / self-interaction options |

2. **Registry.** Backends register per framework under the group **`mace.electrostatics_backends.torch`** (parallel to, not folded into, the kernel-backend group). Discovery **records** import failures (a missing CUDA lib is not fatal); resolution raises. `supports(descriptor) -> bool` is **authoritative**, with coarse `ops`/`devices`/`dtypes`/`supports_double_backward` fields as a cheap first filter.

3. **Op set.** The base dispatched op is the **long-range / k-space sum** `long_range_energy`. The **SCF fixed-point loop lives in the model** (TRN-1 (#1575) model-transform hook) and calls the op per iteration. An **optional span factory** `make_scf_solve(descriptor) -> ScfSolveOp | NotImplemented` lets a backend fuse the whole SCF solve; the model prefers it when present and falls back to the op-by-op path otherwise (the same `NotImplemented` negotiation the kernel-backend span uses).

4. **Reference solver = `graph_electrostatics`** — plain torch, **CPU-capable**, mandatory, the correctness oracle every other solver is checked against. **It consumes the cell the graph carries, and that cell is not always the physical one.** Develop's contract, implemented in `get_neighborhood` (`mace/data/neighborhood.py:7-90`) since #1533 (`050b791a`), has exactly three regimes and the solver must be specified against all three:

   - The **neighbour search always runs against an extended cell** (`neighborhood.py:32-36,62-70`): every non-periodic axis is replaced by `(extent + 2*cutoff + 1) * e_dim`, sized from the *atom extent* along that axis. Extent-based, not origin-dependent — the pre-#1533 `max(abs(positions)) * 5 * cutoff` produced huge boxes that blew PolarMACE's k-space arrays into GPU OOM. Do not reimplement the old sizing.
   - **Any axis periodic → the returned cell is the physical cell** (`:51-58`), because stress normalizes by `det(cell)` and a fabricated vacuum row silently rescales the stress of slabs and mixed-PBC systems. The one exception is a non-periodic axis whose *physical* row is all zeros (a slab built with no vacuum): that row is patched from the extended cell, or `det(cell)=0` NaNs the stress and blows up `rcell`.
   - **Fully aperiodic → the returned cell IS the extended cell** (`:59-60`), deliberately. Stress is meaningless there, and the long-range models need a non-degenerate box: the extended cell is PolarMACE's k-space box and its `det()` is the volume the slab/molecule dipole corrections divide by. Returning a degenerate or identity cell would silently scale those corrections away.

   That cell is a **fourth return value** of `get_neighborhood`, consumed verbatim by `AtomicData.from_config` (`mace/data/atomic_data.py:194,446`), which derives `rcell = 2π·inv(cell)ᵀ` when `|det| > 0` and zeros otherwise (`:410-414`). **The v1 rule is therefore "the solver reads the graph's cell and its `rcell`, and the graph builder owns which cell that is" — not "the solver reads the physical cell".** Adequate vacuum on an aperiodic axis stays the caller's job, exactly as in any 3D-Ewald slab calculation. v1 must reproduce all three regimes and must not silently substitute the physical cell in the aperiodic case.

5. **Selection is build-time config**, resolved once and frozen into the module tree; **nothing dispatches in `forward`**. No runtime setter, no `--electrostatics_backend` flag, no calculator argument — develop has none and v1 must not invent one.

6. **Bit-parity → free backend choice; non-bit-parity → model state.** A solver that is bit-parity with the reference swaps freely at the load boundary. A solver that is **not** bit-parity (the accelerated `nvalchemiops` path needs retraining for transparent replacement) is **model-affecting state**: its identity is serialized with the checkpoint, and a backend **never silently substitutes** a solver that changes numbers. Fallback is a **hard error** for unsupported systems; auto-fallback to the reference is allowed **only** where it is bit-parity with the model's solver.

7. **Force training requires an implicit-differentiation backward.** The reference supplies it. An accelerated inference-only solver declares `supports_double_backward=False` and is **hard-rejected at build time for force/stress training** (loud error), usable for inference. A non-differentiable fixed point silently corrupts training otherwise.

8. **v1.0 scope: reference + dispatch layer only.** `nvalchemiops` is a post-v1.0 follow-on the layer must not foreclose (it should enter as a wheel + a span, zero core edits). Do not add it here.

9. **The extra is created here, and the pin moves with it.** `[electrostatics]` becomes a real extra in the v1 package metadata, carrying the `graph_electrostatics` git pin that today lives in `requirements/polar.txt`. Keep the pin **by commit** — the file's own note says an unpinned git dependency can break CI on its own — and keep the marker name the capability contract already uses (`polar`), so `tests/conftest.py`'s probe and `MACE_REQUIRE_CAPS` keep working unchanged.

**Task:**
1. **`mace_core`**: add `ElectrostaticsSolverDescriptor` and the `BackendCapabilities`-style capability record with `supports(descriptor)` (framework-free, dtype names only, no torch/jax import).
2. **Solver-dispatch layer**: the `mace.electrostatics_backends.torch` entry-point registry (discovery records import failures; resolution raises), `long_range_energy` as the base dispatched op, and the optional `make_scf_solve` span factory with `NotImplemented` fallback. Build-time selection resolved once and frozen; nothing in `forward`.
3. **Reference solver**: wire `graph_electrostatics` into the `mace-torch` experimental module behind the new **`[electrostatics]` extra** (pin by commit, migrated from `requirements/polar.txt`), register it under `mace.electrostatics_backends.torch`, feed it the graph's cell/`rcell` under the three-regime contract of §4, and give it the implicit-differentiation backward required for force training.
4. **Expert training entry point = an `electrostatics` section in the training config, run through plain `mace train`.** No new console script and no new command namespace: the CLI groups by function (`train`, `eval`, `model`, `data`, `export`), not by extension, and this follows the precedent already set for fine-tuning, which is `mace train` with a fine-tune config section rather than its own binary. It matches legacy too — LES training today is `mace_run_train --les_arguments` and polar training is `mace_run_train --model PolarMACE`, both one flag on the ordinary trainer.

   "Expert" means **config-file-only**: the section is absent from `mace train`'s basic-user flag tier and is not surfaced in `mace train --help`, so it is reachable but not advertised. The section may select its own training step, since electrostatics training diverges from the standard loop; that divergence is a value inside the section, not a separate entry point. If the section is present and the `[electrostatics]` extra is not installed, fail with a loud error naming the missing extra — never silently ignore it, and never fall back to the standard step.
5. **CI job**: an **extension-style, paths-filtered** job in the successor of `ci-extensions.yaml` (today's `polar` job is the template: `require-caps: polar`, `markers: not cueq and not gpu`, filter `polar:` over the shared set + `mace/cli/polar_density_cube.py` + `mace/calculators/**` + `tests/extensions/polar/**` + `requirements/polar.txt`), running the conformance + parity tests below.

**Out of scope:** `PolarMACE` on the new architecture, its foundation-model conversion, and the polar user surface — **ELEC-2 (#1592)**. `MACELES` / LES and its `[les]` extra — **ELEC-4 (#1635)**. The split-charge / SCF families (LC-MACE, LSC-MACE, MACE-QEq, FixedPointSCF, EnergyFunctionalSCF) — **ELEC-3 (#1593)**. The accelerated `nvalchemiops` backend — post-v1.0. Deleting the legacy `mace/modules/extensions.py` — RET-4 (#1600).

**Acceptance criteria:**
- [ ] `mace_core`'s `ElectrostaticsSolverDescriptor` + capability record import neither torch nor jax (asserted); `supports(descriptor)` decides periodicity-profile support (e.g. a fully-periodic/z-slab-only solver declines `molecular`/`partial`).
- [ ] The reference solver is discovered via the `mace.electrostatics_backends.torch` entry-point group and resolved **once at build time**; nothing dispatches in `forward`.
- [ ] The three-regime cell contract of §4 is reproduced and asserted: (a) a fully-aperiodic cluster reaches the solver with the **extent-based extended** cell and a finite `det`, and its dipole correction uses that volume; (b) a slab (`pbc=(T,T,F)`) reaches the solver with the **physical** cell, so stress is not rescaled; (c) a zero-vacuum slab has its all-zero row patched so `det(cell) != 0` and `rcell` is finite. No regime silently substitutes another.
- [ ] `gradgradcheck` passes at fp64 on the reference solver's force-training path (implicit-differentiation backward correct).
- [ ] A solver declaring `supports_double_backward=False` is **hard-rejected at build time** for force/stress training (loud error) and usable for inference.
- [ ] A registered non-bit-parity solver that declines an unsupported system produces a **hard error**, never a silent reference substitution.
- [ ] The `electrostatics` config section is absent from `mace train`'s basic flag tier and from `mace train --help`; supplying it without the required extra installed fails with an error naming the extra, rather than being ignored.
- [ ] `[electrostatics]` exists as a real extra pinning `graph_electrostatics` **by commit**; installing the package **without** it leaves the core import clean; the `polar` marker and `MACE_REQUIRE_CAPS=polar` still select the right tests.
- [ ] The paths-filtered CI job runs the electrostatics tests only when those paths change.

**Verify:**
```bash
pip install -e "packages/mace-torch[electrostatics]"
python -c "import mace_core; from mace_core.electrostatics import ElectrostaticsSolverDescriptor; print('descriptor ok')"
python -c "import graph_longrange; print('reference solver import name unchanged')"
python -m pytest packages/mace-torch/tests/electrostatics/test_solver_dispatch.py -v   # registry, build-time resolution, supports()
python -m pytest packages/mace-torch/tests/electrostatics/test_reference_solver.py -v  # gradgradcheck, three-regime cell, capability honesty
python -m pytest packages/mace-torch/tests -m polar                                    # marker + skip-without-extra discipline
```
**Review focus:** the implicit-differentiation backward of the reference solver (`gradgradcheck` on the force path); that the **three-regime** cell contract is reproduced exactly — extended cell for fully-aperiodic systems (this is deliberate, not a bug), physical cell whenever any axis is periodic, zero rows patched — and that neither direction is "fixed" into the other; that solver selection is frozen at build time with nothing resolving in `forward`; and that a non-bit-parity solver is treated as serialized model state and never silently substituted.

---
